test/exec/test_on.cpp

```diff
@@ -0,0 +1,341 @@
+  int recv_value{0};
+  impulse_scheduler sched;
+  auto snd = exec::on(sched, ex::just(13)) | exec::complete_on(inline_scheduler{});
+  auto op = ex::connect(std::move(snd), expect_value_receiver_ex{&recv_value});
+  ex::start(op);
+  // Up until this point, the scheduler didn't start any task; no effect expected
+  CHECK(recv_value == 0);
+
+  // Tell the scheduler to start executing one task
+  sched.start_next();
+  CHECK(recv_value == 13);
+}
+
+TEST_CASE("exec::on calls the given sender when the scheduler dictates", "[adaptors][exec::on]") {
+  bool called{false};
+  auto snd_base = ex::just() | ex::then([&]() -> int {
+    called = true;
+    return 19;
+  });
+
+  int recv_value{0};
+  impulse_scheduler sched;
+  auto snd = exec::on(sched, std::move(snd_base)) | exec::complete_on(inline_scheduler{});
+  auto op = ex::connect(std::move(snd), expect_value_receiver_ex{&recv_value});
+  ex::start(op);
+  // Up until this point, the scheduler didn't start any task
+  // The base sender shouldn't be started
+  CHECK_FALSE(called);
+
+  // Tell the scheduler to start executing one task
+  sched.start_next();
+
+  // Now the base sender is called, and a value is sent to the receiver
+  CHECK(called);
+  CHECK(recv_value == 19);
+}
+
+TEST_CASE("exec::on works when changing threads", "[adaptors][exec::on]") {
+  exec::static_thread_pool pool{2};
+  bool called{false};
+  {
+    // lunch some work on the thread pool
+    ex::sender auto snd = exec::on(pool.get_scheduler(), ex::just()) //
+                          | ex::then([&] { called = true; });
+    ex::start_detached(std::move(snd));
+  }
+  // wait for the work to be executed, with timeout
+  // perform a poor-man's sync
+  // NOTE: it's a shame that the `join` method in static_thread_pool is not public
+  for (int i = 0; i < 1000 && !called; i++)
+    std::this_thread::sleep_for(1ms);
+  // the work should be executed
+  REQUIRE(called);
+}
+
+TEST_CASE("exec::on can be called with rvalue ref scheduler", "[adaptors][exec::on]") {
+  auto snd = exec::on(inline_scheduler{}, ex::just(13));
+  auto op = ex::connect(std::move(snd), expect_value_receiver{13});
+  ex::start(op);
+  // The receiver checks if we receive the right value
+}
+TEST_CASE("exec::on can be called with const ref scheduler", "[adaptors][exec::on]") {
+  const inline_scheduler sched;
+  auto snd = exec::on(sched, ex::just(13));
+  auto op = ex::connect(std::move(snd), expect_value_receiver{13});
+  ex::start(op);
+  // The receiver checks if we receive the right value
+}
+TEST_CASE("exec::on can be called with ref scheduler", "[adaptors][exec::on]") {
+  inline_scheduler sched;
+  auto snd = exec::on(sched, ex::just(13));
+  auto op = ex::connect(std::move(snd), expect_value_receiver{13});
+  ex::start(op);
+  // The receiver checks if we receive the right value
+}
+
+TEST_CASE("exec::on forwards set_error calls", "[adaptors][exec::on]") {
```

Comment:
You don't test an incoming error, only one that the scheduler generates. You also don't test the scheduler transition behaviour if the incoming sender completes with an error. Does it still complete on the receiver scheduler in the case?

Reply:
I've added tests to cover those cases. In short: it still completes on the receiver's scheduler, even when the input sender completes with an error.